File: python/app/processor.py
import uuid


class RetryableProcessorError(Exception):
    pass


class ProcessorDeclinedError(Exception):
    pass


class ProcessorClient:
    def authorize(self, amount: int, currency: str, payment_method_token: str) -> str:
        if amount % 13 == 0:
            raise RetryableProcessorError("temporary processor timeout")
        if amount % 7 == 0:
            raise ProcessorDeclinedError("card declined")
        return f"proc_{uuid.uuid4().hex[:20]}"

File: typescript/src/common/processor.ts
export class RetryableProcessorError extends Error {}
export class ProcessorDeclinedError extends Error {}

export class ProcessorClient {
  authorize(amount: number, paymentMethodToken: string): string {
    if (amount % 13 === 0) {
      throw new RetryableProcessorError("temporary processor timeout");
    }
    if (amount % 7 === 0) {
      throw new ProcessorDeclinedError("card declined");
    }
    return `proc_${paymentMethodToken.slice(0, 6)}`;
  }
}
